process/HOPE-3--Dependency-Injection.md

```diff
@@ -0,0 +1,118 @@
+# HOPE 3 -- Dependency Injection
+
+|             |                                             |
+| ------------| ------------------------------------------- |
+| HOPE:       | 3                                           |
+| Title:      | Dependency Injection                        |
+| Author(s):  | Timothy Crosley <timothy.crosley@gmail.com> |
+| Status:     | Proposed                                    |
+| Type:       | Standard Track                              |
+| Created:    | 25-May-2019                                 |
+| Updated:    | 27-May-2019                                 |
+
+## Abstract
+
+This HOPE proposes adding a robust and universally available dependency injection system to Hug.
+This system would be built with the intention of replacing Hug's current `directive` system, one of the first Python microservice dependency injection systems, solving the shortcomings identified upon extensive usage:
+
+- Directives aren't clearly separated from type annotations.
+- It is impossible to have both a directive and type hint for a single parameter.
+- It isn't easy to swap out directives for full HTTP stack testing (through direct calls make it easy to substitute out).
+- Directives aren't nestable.
+
+## Proposed Dependency Injection System
+
+This HOPE proposes a new dependency injection system with an aim for clarity and reuse.
+
+### Defining a dependency
+
+Defining a dependency will happen via a new hug `provides` decorator:
+
+```python
+    @hug.provides()
+    def mysql_connection(request, config):
+        ...
+```
+
+By default, the name will be the name of the function, but the first argument sent to that directive should enable redefining it:
+
+```python
+    @hug.provides("mysql")
+    def mysql_connection(request, config):
+        ...
+```
+
+It should also be possible to specify different dependency providers for different interfaces:
+
+```python
+    # Multiple interfaces using one provider
+    @hug.provides("mysql", interfaces=['http', 'websocket'])
+    def mysql_web():
+        ...
+
+
+    # For a provider for a single interface, this short hand can be used
+    @hug.cli.provides("mysql")
+    def mysql_cli():
+        ...
+
+
+    # The following would also be a valid way to apply to multiple interfaces
+    @hug.http.provides("mysql")
+    @hug.cli.provides("mysql")
+    def mysql():
+        ...
+```
+
+### Using a dependency
+
+Dependencies will be pulled into individual endpoints via a `requires(` function, that will be called within function/method defaults:
+
+```
+from hug import requires
+
+@hug.http()
+def hello_world(mysql=requires('mysql')):
+    pass
+```
+
+Arguments can be passed to the `requires` function, these will be passed directly to the `provides` function.
```

Comment:
After some thought, I think it makes sense to pass them along, if and only if, the method in question takes no extra optional arguments. Or rather, if the function signature matches that of `functools.partial` it matches an accepted pattern that has almost no chance of name conflicts, while simplifying usage. If it ever goes beyond this, I think you are 100% correct that more explicit usage should be enforced.

Reply:
I think this is a reasonable pattern as long as you can be sure that you will never want to add configuration parameters to the `requires` function. I think it might be worth considering if there's any potential tweaks that might be desired when calling the `requires` function. One example off the top of my head is the idea of potentially renaming a parameter to match a different context:

```python
@hug.provides("mysql")
def mysql_config(host: str, port: int = 3306):
    ...


@hug.cli()
def my_server(
    mysql=requires("mysql", rename={"host": "mysql_host", "port": "mysql_port"}),
    api_host: str,
    api_port: int = 8080,
):
    ...
```
I'm not proposing adding the rename functionality in this HOPE, just highlighting that we'll be preventing such future tweaks if we go with the `functools.partial` pattern. If we're good with making that call, or will have an `advanced_requires` in the future that can be extended I withdraw my concern here.

Reply:
I agree 100% with your assertions here! I think I'm okay limiting the requires endpoint to the very limited subset allowed by the `funtools.partials` signature, and then if other uses are deemed common enough creating other functions for them